crates/sqlexec/src/planner/physical_plan/send_recv.rs

```diff
@@ -0,0 +1,220 @@
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::task::JoinSet;
+
+use super::client_send::ClientExchangeSendExec;
+
+/// Drives execution of the output stream from the server in conjunction with
+/// the send streams to the server.
+///
+/// This exec should produce the final result of a query.
+#[derive(Debug)]
+pub struct SendRecvJoinExec {
+    /// The execution plan producing the output stream.
+    input: Arc<dyn ExecutionPlan>,
+
+    /// The execution plans for sending batches to the remote server.
+    ///
+    /// Note that these only get handled on the the call to the first partition
+    /// execute.
+    send_execs: Arc<Mutex<Vec<ClientExchangeSendExec>>>,
+}
+
+impl SendRecvJoinExec {
+    /// Create a new execution plan that drives both the stream from the input
+    /// execution plan, and the send execs for sending record batches to the
+    /// remote node.
+    ///
+    /// This execution plan should only be create *after* the client send execs
+    /// have been populated by calling `LocalSideTableProvider::scan` (which
+    /// should have already been done by creating the provided input execution
+    /// plan).
+    pub fn new(input: Arc<dyn ExecutionPlan>, refs: Vec<ClientSendExecsRef>) -> SendRecvJoinExec {
+        let send_execs: Vec<_> = refs.into_iter().flat_map(|r| r.take_execs()).collect();
+
+        SendRecvJoinExec {
+            input,
+            send_execs: Arc::new(Mutex::new(send_execs)),
+        }
+    }
+}
+
+impl ExecutionPlan for SendRecvJoinExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> Arc<Schema> {
+        self.input.schema()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.input.output_partitioning()
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(SendRecvJoinExec {
+            input: children[0].clone(),
+            send_execs: self.send_execs.clone(),
+        }))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        // Set up send exec tokio tasks.
+        //
+        // Will be empty if this isn't the first call to execute, which is fine
+        // because we only want these handled once.
+        let send_execs: Vec<ClientExchangeSendExec> =
+            std::mem::take(self.send_execs.lock().as_mut());
+
+        let mut join_set: JoinSet<Result<(), DataFusionError>> = JoinSet::new();
+        for send_exec in send_execs {
+            let context = context.clone();
+            join_set.spawn(async move {
+                let mut stream = send_exec.execute(0, context)?;
```

Comment:
We "know" here that a `send_exec` has one output partition. Would this be a potentially dangerous assumption? 🤔

Reply:
In `ClientExchangeSendExec`, is it really safe to assume it's only one partition? There it seems even more doubtful.
```
98: let input = self.input.execute(0, context)?;
```

Reply:
It's currently assumed that we're only working with a single partition across `ClientExchangeSendExec` and `ClientExchangeRecvExec`: https://github.com/GlareDB/glaredb/blob/main/crates/sqlexec/src/planner/physical_plan/client_send.rs#L85.

We could pretty easily extend this though to handle more partitions. Instead of of creating one tokio task per send exec, we would create one _per partition_ per send exec. I believe the code changes would be minimal for this. `ClientExchangeRecvExec` would need to be tweaked to take in a vector of broadcast ids and pull from all streams during execution. Right now it just assumes a single partition (using a single broadcast id).